src/Microsoft.Azure.WebJobs.Host/JobHost.cs

```diff
@@ -329,37 +335,92 @@ private static IFunctionDefinition ResolveFunctionDefinition(MethodInfo method,
             return function;
         }
 
-        private async Task<JobHostContext> CreateContextAndLogHostStartedAsync(CancellationToken cancellationToken)
+        private void ThrowIfDisposed()
         {
-            JobHostContext context = await _config.CreateAndLogHostStartedAsync(this, _shutdownTokenSource.Token, cancellationToken);
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(null);
+            }
+        }
 
-            lock (_contextLock)
+        // If multiple threads call this, only one should do the init. The rest should wait
+        // When this task is signalled, _context is initialized. 
+        private Task EnsureHostStartedAsync(CancellationToken cancellationToken)
+        {
+            if (_context != null)
             {
-                if (_context == null)
+                return Task.CompletedTask;
+            }
+
+            TaskCompletionSource<bool> tsc = null;
+
+            lock (_stopTaskLock)
+            {
+                if (_initializationRunning == null)
                 {
-                    _context = context;
-                    _listener = context.Listener;
+                    // This thread wins the race and owns initialing. 
+                    tsc = new TaskCompletionSource<bool>();
+                    _initializationRunning = tsc.Task;
                 }
             }
 
-            _logger = _context.LoggerFactory?.CreateLogger(LogCategories.Startup);
+            if (tsc != null)
+            {
+                // Ignore the return value and use tsc so that all threads are awaiting the same thing. 
+                Task ignore = RuntimeInitAsync(cancellationToken, tsc);
+            }
 
-            return _context;
+            return _initializationRunning;
         }
 
-        private Task EnsureHostStartedAsync(CancellationToken cancellationToken)
+        // Caller gaurantees this is single-threaded. 
+        // Set initializationTask when complete, many threads can wait on that. 
+        // When complete, the fields should be initialized to allow runtime usage. 
+        private async Task RuntimeInitAsync(CancellationToken cancellationToken, TaskCompletionSource<bool> initializationTask)
```

Comment:
We have  2 categories of services: I want to distinguish between the phase 1 (Static) vs. phase 2(runtime). 

---
In reply to: [126260074](https://github.com/Azure/azure-webjobs-sdk/pull/1202#discussion_r126260074) [](ancestors = 126260074)